src/sp/it/pl/util/file/Util.kt

```diff
@@ -13,53 +13,49 @@ import kotlin.streams.asStream
 import kotlin.streams.toList
 
 /**
- * For files 'filename.extension' is returned.
- * For directories only name is returned.
- * Root directory returns 'X:\' string.
- *
- * Use instead of [File.getName] which returns empty string for root directories.
+ * Returns the [File.getName] or [File.toString] if this is the root directory,
+ * since that returns an empty string otherwise.
  *
  * @return name of the file with suffix
  */
-val File.nameOrRoot: String get() = name.takeUnless { it.isEmpty() } ?: toString()
+val File.nameOrRoot: String
+    get() = name.takeUnless { it.isEmpty() } ?: toString()
 
 /**
- * For files name with no extension is returned.
- * For directories name is returned.
- * Root directory returns 'X:\' string.
+ * Returns the [File.nameWithoutExtension] or [File.nameOrRoot] if this is a directory.
  *
- * @return name of the file without suffix
+ * @return name of the file without extension
  */
-// TODO: does not work for directories with '.'
-val File.nameWithoutExtensionOrRoot: String get() = nameWithoutExtension.takeUnless { it.isEmpty() } ?: toString()
+val File.nameWithoutExtensionOrRoot: String
+    get() = if(isDirectory) nameWithoutExtension else nameOrRoot
```

Comment:
It doesn't break existing functionality, it fixes it. I prefer safety over ms shaved off. As said, every call that doesn't require that check and is sure to only point to files (or even directories that are not the root) should simply use `nameWithoutExtension` directly. But as far as I saw, this method is mostly called for visual representation. Why would you use such a method in performance-critical parts anyways? There you likely don't even care whether the root is an empty string or not...

Reply:
1 It is not ms, it is even dozens of seconds! I know because it happened to me in this very application, which is why I started to aggressively remove every isFile/isDirectory check I could. This is not a potential performance problem, this is guaranteed one. Listing over 20000 files, which is a common use case in this application for me goes from 2 seconds to 20 and it completely destroys the UX.

2 Indeed this is only used for visual representation (although the intention is not clear from the design/javadoc) and thus there is no safety problem here. This does have to be fixed, but this not acceptable solution.

3 Why would you use such a method in performance-critical parts anyways Because it was designed to be performant. Unfortunately the . problem was not apparent for a long time and the apparent solution is not viable anymore. Fixing this requires going over every use case and devising a counter measure.

4 The methods exists exactly because I care about whether the file represents a root. Java returns empty string for root, which is unacceptable for any UI case. As you can see theres a conflict of interest here. The method is intended for UI both in functionality and speed, but your solution takes that away.

It is not like I wish to live with bugs in the application, but this is a complex issue. I already said that I was even thinking about some kind of I/O monad to prevent blocking calls on UI, namely from calls such as isFile() check. I also thought about making a FFile and DFile subclasses for File to prevent misuse of some of the File methods. You just demonstrated how easy it is to silently put isFile() somewhere with no regard to the implications.

Reply:
PS: very clever solution which I already experimented with is to use File subclass that caches the result of its isFile() call. The danger here is that the isDirectory is not simply inverse of isFile() and that isFile()/isDirectory() invoke exists() check, which may not hold true in some later time, which the cached value would not reflect.

Reply:
I'll reverse it for now and we'll look into that separately.

Reply:
Yes. It absolutely has to be taken care of.